torch/distributed/distributed_c10d.py

```diff
@@ -1714,6 +1714,10 @@ def broadcast_object_list(object_list, src=0, group=None):
         src (int): Source rank from which to broadcast ``object_list``.
         group: (ProcessGroup, optional): The process group to work on. If None,
             the default process group will be used. Default is ``None``.
+        dist_device (``torch.device``, optional): If not None, then intermediate
+            tensor representations of objects, or any tensors within the contained
```

Comment:
> "or any tensors within the contained objects, will be moved to this device before broadcasting"

Sorry for missing this before. Upon closer examination, I do not believe this is true. The size and object tensors are indeed moved to `dist_device` before broadcasting, but those two tensors are the result of serializing the inputted objects. I do not think that given the current implementation of `_object_to_tensor()` and `_tensor_to_object()`, the tensors contained in the object pre-serialization will be moved to `dist_device`.

The `dist_device` argument is just the device used for the broadcast. Hence, I think precisely, this PR simply adds an argument that allows a process to broadcast from a specified device. 
Before:
- If the backend is NCCL, then the broadcast uses `torch.cuda.current_device()`.
- If the backend is not NCCL, then it uses `torch.device("cpu")`.

Now:
- If the backend is NCCL, a process can use something other than `torch.cuda.current_device()` (but that something must still be a GPU).
- If the backend is not NCCL, a process can use something other than CPU. 

Given this realization, I wonder how useful this change is. @rohan-varma @mrshenli 
- If the backend is NCCL, why not use `torch.cuda.current_device()`? Even if there exists a second GPU available, why would that be any better to use? 
- If the backend is not NCCL, then would be using GPU for the broadcast be desirable? I think for both GLOO and MPI backends, using CPU is preferred (correct me if I am wrong).

Reply:
I think it's useful to have the flexibility when backend is NCCL to control which cuda device tensors are placed on. It makes it more similar to `dist.broadcast(tensor)` APIs where the tensor can be on a device which is not necessarily the current device. This use case might be rare but I don't see a disadvantage to enabling it.

Regarding Gloo backend, Gloo is also frequently used with GPU, although for this particular API, I suspect perf would be better if the user simply used the CPU. But I'm not sure if that's something we should enforce and it would make the API more prescriptive. Given that I think it's okay to respect the passed in device being a GPU for gloo backend, if CUDA is enabled.

Reply:
Given rohan's comment, WDYT Andrew? Sorry I was not quite familiar with the code, will defer to both of you :)

Reply:
I think what @rohan-varma said makes sense. Let us move forward with adding this `device` argument, which specifies which device the size and object tensors are moved to before broadcasting. 

In my follow-up PR (which should be ready by tomorrow), I plan to introduce a `map_location` argument, which specifies which device the tensors in the broadcasted object will be loaded to upon being received.

Reply:
SG

Reply:
Minor comment:

> or any tensors within the contained objects, will be moved to this device before broadcasting.

This wording in the comment may still be misleading since it glosses over the serialization step. At least to me, the comment suggests that the tensors contained in the objects are moved to `device` and are themselves broadcasted; however, actually the object is serialized and converted to a tensor. That tensor is moved to `device` and broadcasted.

Reply:
updated per your last sentence, how does it sound?